src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,502 @@
+    constructor(
+        IDelegationManager _delegationManager,
+        IStrategyManager _strategyManager,
+        uint32 _maxPaymentDuration,
+        uint32 _MAX_RETROACTIVE_LENGTH,
+        uint32 _MAX_FUTURE_LENGTH,
+        uint32 _GENESIS_PAYMENT_TIMESTAMP
+    )
+        PaymentCoordinatorStorage(
+            _delegationManager,
+            _strategyManager,
+            _maxPaymentDuration,
+            _MAX_RETROACTIVE_LENGTH,
+            _MAX_FUTURE_LENGTH,
+            _GENESIS_PAYMENT_TIMESTAMP
+        )
+    {
+        _disableInitializers();
+        ORIGINAL_CHAIN_ID = block.chainid;
+    }
+
+    /**
+     * @dev Initializes the addresses of the initial owner, pauser registry, paymentUpdater and
+     * configures the initial paused status, activationDelay, and globalOperatorCommissionBips.
+     */
+    function initialize(
+        address initialOwner,
+        IPauserRegistry _pauserRegistry,
+        uint256 initialPausedStatus,
+        address _paymentUpdater,
+        uint32 _activationDelay,
+        uint32 _calculationIntervalSeconds,
+        uint16 _globalCommissionBips
+    ) external initializer {
+        _DOMAIN_SEPARATOR = _calculateDomainSeparator();
+        _initializePauser(_pauserRegistry, initialPausedStatus);
+        _transferOwnership(initialOwner);
+        _setPaymentUpdater(_paymentUpdater);
+        _setActivationDelay(_activationDelay);
+        _setCalculationIntervalSeconds(_calculationIntervalSeconds);
+        _setGlobalOperatorCommission(_globalCommissionBips);
+    }
+
+    /**
+     * @notice Creates a new range payment on behalf of an AVS, to be split amongst the
+     * set of stakers delegated to operators who are registered to the `avs`
+     * @param rangePayments The range payments being created
+     * @dev Expected to be called by the ServiceManager of the AVS on behalf of which the payment is being made
+     * @dev The duration of the `rangePayment` cannot exceed `MAX_PAYMENT_DURATION`
+     * @dev The tokens are sent to the `PaymentCoordinator` contract
+     * @dev Strategies must be in ascending order of addresses to check for duplicates
+     * @dev This function will revert if the `rangePayment` is malformed,
+     * e.g. if the `strategies` and `weights` arrays are of non-equal lengths
+     */
+    function payForRange(
+        RangePayment[] calldata rangePayments
+    ) external onlyWhenNotPaused(PAUSED_PAY_FOR_RANGE) nonReentrant {
+        for (uint256 i = 0; i < rangePayments.length; i++) {
+            RangePayment calldata rangePayment = rangePayments[i];
+            bytes32 rangePaymentHash = keccak256(abi.encode(msg.sender, paymentNonce, rangePayment));
+            require(
+                !isRangePaymentHash[msg.sender][rangePaymentHash],
+                "PaymentCoordinator._payForRange: range payment hash already submitted"
+            );
+            isRangePaymentHash[msg.sender][rangePaymentHash] = true;
+
+            _payForRange(rangePayment);
```

Comment:
Not a huge fan of calling this `_payForRange`. I think this method does two things, neither of which I would call `_payForRange`:

`_validateRangePayment`

and

`_depositRangePayment`

Could we either rename this, or split it into 2 methods?

Reply:
have tried implementing this here: https://github.com/Layr-Labs/eigenlayer-contracts/pull/530/files

Reply:
Thanks for this 👍